src/misc/KeyManager.js

```diff
@@ -99,114 +119,106 @@ class KeyManager {
 	}
 
-	_getShortcutName(shortcut: Shortcut): string {
-		return ((shortcut.meta) ? Keys.META.name + " + " : "")
-			+ ((shortcut.ctrl) ? Keys.CTRL.name + " + " : "")
-			+ ((shortcut.shift) ? Keys.SHIFT.name + " + " : "")
-			+ ((shortcut.alt) ? Keys.ALT.name + " + " : "")
-			+ shortcut.key.name
+	_handleKeydown(e: KeyboardEvent): void {
+		let keyCode = e.which
+		let keysToShortcuts = (this._modalShortcuts.length > 1)
+			? this._keyToModalShortcut
+			: this._keyToShortcut
+		let shortcut = keysToShortcuts.get(this._createKeyIdentifier(keyCode, e.ctrlKey, e.altKey, e.shiftKey, e.metaKey))
+		if (shortcut != null && (shortcut.enabled == null || shortcut.enabled())) {
+			if (shortcut.exec({
+				keyCode,
+				key: e.key,
+				ctrl: e.ctrlKey,
+				alt: e.altKey,
+				shift: e.shiftKey,
+				meta: e.metaKey
+			}) !== true) {
+				e.preventDefault()
+			}
+		}
 	}
 
 	_createKeyIdentifier(keycode: number, ctrl: ?boolean, alt: ?boolean, shift: ?boolean, meta: ?boolean): string {
 		return keycode + (ctrl ? "C" : "") + (alt ? "A" : "") + (shift ? "S" : "") + (meta ? "M" : "")
 	}
 
 	openF1Help() {
-		Promise.all([import("../gui/base/Dialog.js"), import("../gui/base/ButtonN"), import("../gui/base/TextFieldN")])
-		       .then(([{Dialog}, {ButtonType}, {TextFieldN}]) => {
-			       if (this._helpDialog && this._helpDialog.visible) {
-				       return
-			       }
-			       let shortcuts = ((this._modalShortcuts.length > 1)
-				       ? this._modalShortcuts
-				       : Array.from(this._keyToShortcut.values()).concat(this._desktopShortcuts)) // we do not want to show a dialog with the shortcuts of the help dialog
-			       let textFieldAttrs = shortcuts.filter(shortcut => shortcut.enabled == null || shortcut.enabled())
-			                                     .map(shortcut => {
-				                                     return {
-					                                     label: () => this._getShortcutName(shortcut),
-					                                     value: stream(lang.get(shortcut.help)),
-					                                     disabled: true
-				                                     }
-			                                     })
-			       this._helpDialog = Dialog.largeDialog({
-				       left: [{label: 'close_alt', click: () => neverNull(this._helpDialog).close(), type: ButtonType.Secondary}],
-				       middle: () => lang.get("keyboardShortcuts_title")
-			       }, {
-				       view: () => {
-					       return m("div.pb", textFieldAttrs.map(t => m(TextFieldN, t)))
-				       }
-			       }).addShortcut({
-				       key: Keys.ESC,
-				       exec: () => neverNull(this._helpDialog).close(),
-				       help: "close_alt"
-			       }).show()
-		       })
+		// first, decide if we're showing the modal shortcuts or the standard shortcuts
+		// the help dialog will register its own shortcuts which would override the
+		// standard shortcuts if we did this lazily
+		this._showModalShortcuts = this._modalShortcuts.length > 1
```

Comment:
Why do we do this here? In `_handleBatch` we check the value of `_showModalShortcuts`, then we set `_dialogAttrs` accordingly. Can we not just check `_modalShortcuts.length` in `_handleBatch` and handle it there?

Reply:
added a better explanation